testing-styleguide/README.md

```diff
@@ -0,0 +1,496 @@
+
+#### 🛠️ Service layer
+
+Using the same logic as in the Manager layer, we mock the dependencies and test if they are called appropriately.
+
+Let's follow the example of the `HiringUpdater`. In order to update the ActiveRecord records, it needs to first validate the parameters received from the manager via a `HiringUpdateValidator`. Then it either raises an Exception if parameters are invalid or proceeds to call a `DelegateRepository` in order to effectively persist the updates to the database. After updating, it returns the updated Hiring.
+
+❌ Bad
+
+```ruby
+context 'when service is called with valid params' do
+  it 'updates the hiring' do
+    # ... code asserting if hiring is updated
+    # !! The repository that updates jobs already tests this
+  end
+end
+```
+
+✅ Good
+
+```ruby
+let(:provide_hiring_repo) { instance_double Repositories::Hirings::DelegateRepository }
+let(:hiring_update_validator) { instance_double Validators::Hiring::HiringUpdateValidator }
+
+let(:hiring) { build_stubbed(:hiring) }
+
+context 'when service is called with valid params' do
+  let(:params) do
+    {
+      id: hiring.id,
+      status: Hiring.status.approved
+    }
+  end
+
+  before do
+    allow(subject).to receive(:provide_hiring_repo).and_return provide_hiring_repo
+    allow(Validators::Hiring::HiringUpdateValidator).to receive(:new).with(kind_of(Hash), kind_of(Symbol)).and_return hiring_update_validator
+
+    allow(hiring_update_validator).to receive(:valid?).and_return true
+    allow(provide_hiring_repo).to receive(:update).with(kind_of(Hash))
+
+    subject.update(params)
+  end
+
+  it "should call the hiring repository with appropriate arguments the right number of times" do
+    expect(provide_hiring_repo).to have_received(:update).with(expected_params).exactly(1).times
+  end
+
+end
+```
+
+Here, the same performance consideration applies: prefer `FactoryBot.build_stubbed` over `FactoryBot.create` whenever possible.
+
+#### 🚫 Validator layer
+
+The validator layer is responsible to check the data that is been passed to the database.
+Here we want to assert if the validator is setting the errors properly according to the parameters validation rules.
+
+Always cover all edge cases for possible types as this can avoid a lot of validation problems, which won't be necessarily tested in outer players.
+
+```ruby
+module Validators
+  module Candidate
+    class CandidateUpdateValidator < Validator
+      def initialize(attrs = {}, scope = :update)
+        super attrs, scope
+      end
+
+      private
+
+      attr_reader :available, :name
+
+      validates :available, inclusion: [true, false], if: -> { scope == :update }
+      validates :name, presence: true, allow_blank: false, if: -> { scope == :update }
+    end
+  end
+end
+```
+❌ Bad
```

Comment:
In the other examples you state why the Bad test is bad an why the Good is good. I miss the same here in this example

Reply:
thanks! will do !